content/ADR165-resolving-component-numbers.md

```diff
@@ -0,0 +1,155 @@
+
+And due to the imperative nature of the code, defining components (and their schemas) makes it difficult to statically analyze the environment to extract information from it to create better tools. It would be highly beneficial to "analyze" a project to assess its components rather than "run" the code to extract its information.
+
+## Solution space exploration
+
+The solution space exploration will evaluate three big problems, **defining the component number**, **coding ergonomics** and **static reflection**.
+
+By **defining the component number**, the stability and speed of the internal protocols will be the key factor. Among with using third party libraries without component collisions.
+
+With **coding ergonomics**, how the developers _define_ and _use_ the components will be evaluated.
+
+And lastly, **static reflection** contemplates the ability of the approach to extract information from the raw assets of the project (and its libraries) to enable better creation tools and static representations of the scenes and composites.
+
+### Ideal solution
+
+The ideal solution is a _nominal_ collection of components, in which _names_ of components are used to identify the schemas.
+
+Characteristics:
+- Developers will never see the component number under normal circumstances.
+- The fully qualified name (FQN), defined as `package-name/route/to/component/Component` is used to derivate the deterministic component number.
+- Component numbers have no collisions, a Hash algorithm without collisions may be used to reduce the domain of the `Hash(FQN) -> integer`.
+- Components are imported from the files that defined them, using native TypeScript imports.
+- The tooling is able to traverse and analyze all the component definitions, extract their FQN and schema, in order to produce static information to render the editor UI and text (de)serializers.
+
+#### Calculating the component number
+
+For the rest of the document, we can assume that the component number will be derived using a hash of the FQN. The caveat is that the image of the function will be limited to `int32`, and that greatly increases the chances of collisions. This will mostly cover the **defining the component number** dimension, focusing the domain of the problem to how to define the FQN.
+
+For the sake of simplicity in implementation and distribution of hashes, a regular CRC32 will be used to calculate the component number using a function like this:
+
+```ts
+// components with fixed numbers, used for the rendering for performance
+// reasons
+const MAX_RESERVED_COMPONENT_NUMBER = 2048
+// NOTE: this function will always return uint32
+function computeComponentNumber(fullyQualifiedName: string): number {
+  return (unsignedCRC32(fullyQualifiedName) + 2048) & 0xFFFF_FFFF
+}
+```
+
+### Alternative 1
+
+Using a convention over configuration approach, components are defined in a well-known folder, using a constrained type to prevent imperative code and race conditions to happen.
```

Comment:
```suggestion
Using a convention over configuration approach, components are defined in a well-known folder, using a constrained type to prevent imperative code and race conditions.
```